doc/dev/kdb-operations.md

```diff
@@ -0,0 +1,159 @@
+7. Configure `KDB` instance with real mountpoints
+8. Add hardcoded mountpoints to `KDB` instance
+
+Namespaces in mountpoint configs:
+
+- cascading mountpoints act as an alias for identical `proc:/`, `dir:/`, `user:/`, `system:/` and `default:/` mountpoints.
+- `meta:/` mountpoints are illegal
+- explicit `default:/` mountpoints are illegal
+  They can only exist via cascading mountpoints.
+  This is so that default values can be checked and validated by plugins.
+- `spec:/` mountpoints may exist, but they are treated differently during `get` and `set`
+- `proc:/` mountpoints may exist, but they are read-only (see below)
+- `dir:/`, `user:/` and `system:/` mountpoints can be created without restrictions
+
+Other restrictions:
+
+- Creating mountpoints for root keys (`system:/`, `user:/`, etc.) is not allowed.
+  These parts of the KDB always use hardcoded backends that can only be configured at compile-time.
+- Creating a mountpoint for `system:/elektra` or any mountpoint below is not allowed.
+  This part of the KDB will always be stored in the bootstrap backend.
+  While the only part that is actually required for the bootstrap process is the one below `system:/elektra/mountpoints`, everything below `system:/elektra` is reserved and should only be used internally.
+  Additionally, `system:/elektra/version` and `system:/elektra/modules` will always point to hardcoded read-only backends containing information about this Elektra installation.
+
+## `get` Operation
+
+The purpose of the `get` operation is to read data stored in backends into a `KDB` instance.
+
+> **Note:** Some details of a `get` operation are defined in the [contract with backend plugins](./backend-plugins.md).
+
+Properties of `kdbGet()`:
+
+- After calling `kdbGet (kdb, ks, parentKey)`, the KeySet `ks` will contain _all keys_ (including their values) that are stored in _any backend_ with a mountpoint that is _below `parentKey`_.
+- After calling `kdbGet (kdb, ks, parentKey)`, below `parentKey` the KeySet `ks` will _only_ contain keys that are stored in a backend.
+- The KeySet `ks` may contain other keys not below `parentKey`.
+  These keys fall into one of two categories:
+  1. Keys that are stored in a backend that is not below `parentKey`.
+     `kdbGet()` may decide that it is more efficient (e.g. because of a cache) to return more keys than requested.
+  2. Keys that were already present in `ks` when `kdbGet()` was called.
+
+In simpler terms, this means to the caller it looks as if `kdbGet()` had removed all keys below `parentKey` from `ks` and then loaded the data from the backends.
+Below `parentKey` the KeySet `ks` correctly represents the state of the KDB.
+For the rest of `ks` there are no such guarantees.
+
+TODO: cache correct? notifications? (just call at the end?)
+
+The basic flow of this operation is:
+
+1. Determine the backends needed to read all keys below `parentKey`
+2. Run the `init` phase on all the backends that haven't been initialized
```

Comment:
Do you mean `elektra<Plugin>Open`, i.e. the `kdbOpen` counterpart of plugins?

`elektra<Plugin>Open` is already called during `kdbOpen()` (i.e. the `open` operation). This `init` phase is deferred until `get`, because we don't need to initialize all mountpoints. Only those below the `parentKey` we use are needed.

For example if we mounted `/foo` and `/mybackend` and only call `kdbGet()` with `/mybackend` (because `/foo` belongs to another application), it would be unnecessary to initialize `/foo` as well.

Reply:
> Do you mean elektra<Plugin>Open, i.e. the kdbOpen counterpart of plugins?

Yes.

> For example

The example was very helpful, please add more of them into the docu!

Reply:
> The example was very helpful, please add more of them into the docu!

I will add a separate section with explanations of why we do things this way. I think this makes more sense than constantly putting examples and explanations in the middle.

If somebody wants to write a new Rust Version of `libelektra-kdb` they don't care _why_ we do it this way. They want to know exactly what code they need to write. That's also why it's written like this. I wrote this basically as instructions to myself ;)

Reply:
Very good idea! It is elegant to have the spec + rationale separated. :sparkling_heart: